### DISCUSSION TITLE
SED not receiving CoAP responses?

### DISCUSSION THREAD

**Discussion by sjlongland** [openthread/openthread#8589] [Q&A] [answered]
Hi all,

I'm busy trying to coax a CC2538-based platform to operate in a sleepy end device role for an upcoming battery-operated pulse counter.  The SED sends CoAP requests, then goes to sleep (for 50ms at a time), but then seems to miss the CoAP response.

I've built OpenThread for MTD (presently on commit `2db38b033`).  My code is based on an example Sleepy End Device firmware for the EFR32MG12 microcontrollers.  I have the CC2538 successfully going into PM1 deep sleep (sadly, can't do PM2 because OpenThread built with JOINER=1 consumes ~18kiB RAM and the CC2538 in PM2 only has 16kiB available due to the lower 16kiB being "switched off" in that mode).

Notably in the EFR32 example:

* `otLinkSetPollPeriod` is called to set the polling period: https://github.com/openthread/ot-efr32/blob/main/examples/sleepy-demo/sleepy-mtd.c#L81
* `otLinkSetLinkMode` is used to tune some further parameters: https://github.com/openthread/ot-efr32/blob/main/examples/sleepy-demo/sleepy-mtd.c#L83-L86

In previous examples, the state change callback was also hooked to inhibit sleeping when the SED was not connected to a parent node (can't find it in the current version of the code):

* https://github.com/openthread/openthread/blob/thread-reference-20200818/examples/platforms/efr32mg12/sleepy-demo/sleepy-demo-mtd/main.c#L104
* https://github.com/openthread/openthread/blob/thread-reference-20200818/examples/platforms/efr32mg12/sleepy-demo/sleepy-demo-mtd/main.c#L195-L223

I've since amended code that monitors the radio state:

```
#ifdef WIDESKYWM
/*!
 * Determine whether it is safe for us to go to sleep yet.
 */
static _Bool wswm_can_sleep() {
	/* Are we attached to a network? */
	if (!allow_deep_sleep) // aka sAllowDeepSleep in the EFR32 demos
		return false;

	/* Is the radio asleep? */
	switch (otPlatRadioGetState(openthread_instance)) {
	case OT_RADIO_STATE_RECEIVE:
	case OT_RADIO_STATE_TRANSMIT:
	case OT_RADIO_STATE_INVALID:
		return false;
	default:
		break;
	}

	// more internal application checks here

	return true;
}
#endif
```

I'm assuming that, if the radio is indeed asleep, that it's safe for the CPU to sleep too.

I'm using the CC2538's sleep timer to "advance" the alarm timer when coming out of sleep mode.  I have contemplated whether I drop the use of the `SYSTICK` timer altogether and just lean on the 32kHz sleep timer instead, since it gives me a 32-bit unsigned integer field (annoyingly, in 8-bit chunks; I'm guessing it's a 8051 peripheral grafted onto an ARM system bus), so time-keeping there _should_ be taken care of.

I also note the notification regarding `otLinkSetPollPeriod` and SEDs; so I tried replacing that with `otThreadSetChildTimeout` (and converting the argument to seconds).  No luck there.  I had a look at the underlying code for this:

* `otThread{Get,Set}ChildTimeout` (https://github.com/openthread/openthread/blob/main/src/core/api/thread_api.cpp#L47-L55) are wrappers around the `Mle` `{Get,Set}Timeout` method (https://github.com/openthread/openthread/blob/main/src/core/thread/mle.hpp#L477-L491)
* `Mle::{Get,Set}Timeout` are getters/setters for the `mTimeout` property, which is initialised here: https://github.com/openthread/openthread/blob/main/src/core/thread/mle.cpp#L96
* `kMleEndDeviceTimeout` is defined here: https://github.com/openthread/openthread/blob/main/src/core/thread/mle_types.hpp#L178
* `OPENTHREAD_CONFIG_MLE_CHILD_TIMEOUT_DEFAULT` is defined with a default value of 240(!) here: https://github.com/openthread/openthread/blob/main/src/core/config/mle.h#L61-L69

The default value of `otThreadGetChildTimeout` makes me think I'm possibly barking up the wrong tree here.

I haven't yet isolated whether the FTDs on the network (one Fanstel USB840X; nRF52840-based with RCP and OTBR on a RevolutionPi Core 3, and three CC2538s running FTD firmware) are failing to pass on the replies to the SED for some reason, or whether the SED is failing to hear them.

I see the SED periodically "wake up" every few seconds and send a packet (the current consumption spikes to ~230mA -- there's a +22dBm PA), so it _does_ appear to be polling for data, perhaps from its parent.

----

* Is there something I've missed that tells the parent "I'm going to sleep now", or some specific API call that tells my code "it is safe to sleep now"?
* Is it correct to be setting the child timeout to influence the sleep period (given its default is a very lengthy 4 minutes)?

**Comment by sjlongland** [comment]
This is a packet capture showing the incoming request and the subsequent response, the latter of which does not seem to be received by the device.

```
Internet Protocol Version 6, Src: fd34:fe56:7891:10:8c69:ab6c:261:8be5, Dst: fd9e:e308:e14c:b877::fd
User Datagram Protocol, Src Port: 5683, Dst Port: 61616
Constrained Application Protocol, Confirmable, POST, MID:63582
    01.. .... = Version: 1
    ..00 .... = Type: Confirmable (0)
    .... 1000 = Token Length: 8
    Code: POST (2)
    Message ID: 63582
    Token: 98061a52fa50ad50
    Opt Name: #1: Uri-Path: r
        Opt Desc: Type 11, Critical, Unsafe
        1011 .... = Opt Delta: 11
        .... 0001 = Opt Length: 1
        Uri-Path: r
    Opt Name: #2: Uri-Path: 00124b001cddb919
        Opt Desc: Type 11, Critical, Unsafe
        0000 .... = Opt Delta: 0
        .... 1101 = Opt Length: 13
        Opt Length extended: 3
        Uri-Path: 00124b001cddb919
    [Uri-Path: /r/00124b001cddb919]

Frame 53: 129 bytes on wire (1032 bits), 129 bytes captured (1032 bits) on interface 0
Linux cooked capture
Internet Protocol Version 6, Src: fd9e:e308:e14c:b877::fd, Dst: fd34:fe56:7891:10:8c69:ab6c:261:8be5
User Datagram Protocol, Src Port: 61616, Dst Port: 5683
Constrained Application Protocol, Confirmable, 2.05 Content, MID:65142
    01.. .... = Version: 1
    ..00 .... = Type: Confirmable (0)
    .... 1000 = Token Length: 8
    Code: 2.05 Content (69)
    Message ID: 65142
    Token: 98061a52fa50ad50
    Opt Name: #1: Content-Format: application/cbor
        Opt Desc: Type 12, Elective, Safe
        1100 .... = Opt Delta: 12
        .... 0001 = Opt Length: 1
        Content-type: application/cbor
    End of options marker: 255
    [Request In: 51]
    [Response Time: 3.050908450 seconds]
    [Uri-Path: /r/00124b001cddb919]
    Payload: Payload Content-Format: application/cbor, Length: 50
```

Yet, it can receive ICMPv6 traffic:

```
root@RevPi37442:~# ping6 -n fd34:fe56:7891:10:8c69:ab6c:261:8be5
PING fd34:fe56:7891:10:8c69:ab6c:261:8be5(fd34:fe56:7891:10:8c69:ab6c:261:8be5) 56 data bytes
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=1 ttl=63 time=2473 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=2 ttl=63 time=1681 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=3 ttl=63 time=851 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=4 ttl=63 time=2224 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=5 ttl=63 time=1586 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=6 ttl=63 time=709 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=7 ttl=63 time=1981 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=8 ttl=63 time=1420 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=9 ttl=63 time=2405 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=10 ttl=63 time=1619 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=11 ttl=63 time=793 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=12 ttl=63 time=4002 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=13 ttl=63 time=2935 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=14 ttl=63 time=2015 ms
64 bytes from fd34:fe56:7891:10:8c69:ab6c:261:8be5: icmp_seq=15 ttl=63 time=1155 ms
^C
--- fd34:fe56:7891:10:8c69:ab6c:261:8be5 ping statistics ---
31 packets transmitted, 15 received, 51.6129% packet loss, time 47ms
rtt min/avg/max/mdev = 708.795/1856.578/4002.129/853.403 ms, pipe 4
```

**Comment by sjlongland** [comment, marked as answer]
So, further investigation… I was able to do further debugging on a similar device that actually had a functional serial console, so was able to debug this a little more fully.

The problem (which was hard to determine from "outside") was that the device actually froze due to bugs in my hardware abstraction layer -- but with no UI to interact with, that was hard to see.

I ended up reverting my changes that called `otThreadSetChildTimeout` since that caused more problems than it solved.  The documentation in this area is confusing I think.

More recent EFR32 demos set `sAllowDeepSleep` based on the state of a push-button on the hardware (code reads the push-button, and toggles `sAllowDeepSleep` if pressed), and no longer make the Thread node state a dependency -- so that test is now removed from my code.  I'm getting better results with the following:

* At power-on time:
   * Set the `otThreadSetChildTimeout` to some smallish value (I am using 5000ms right now)
   * Use `otThreadSetLinkMode` to set `mRxOnWhenIdle` to `false` so the radio gets turned off on idle
* In the main loop:
   * Check `otPlatRadioGetState` returns a state other than `OT_RADIO_STATE_RECEIVE`, `OT_RADIO_STATE_TRANSMIT` or `OT_RADIO_STATE_INVALID` before going to sleep

I find also if I delay my initial sleep until `otThreadGetDeviceRole` returns `OT_DEVICE_ROLE_CHILD`, I get more reliable mesh connection -- at the moment I wait a maximum of 5 seconds before going to sleep anyway, but I'd like to try and get rid of this test since I don't want to be draining the battery if there's no parent around -- but this would be a separate matter not directly related to this discussion, so I'll put this to bed for now.

**Reply by jwhui** [reply]
Glad to hear that you were able to make progress on this.

In theory, as long as OpenThread has no tasklets to run, you the processor should be able to "sleep" until the next interrupt (e.g. alarm, radio, etc.). Of course, the level of "sleep" depends on the specific hardware capabilities.

Regarding Child Timeout vs. Poll Period - the Child Timeout tells the Parent how long after not hearing from the Child before removing the corresponding entry from its child table. The Poll Period tells the Child how often to poll the Parent for any buffered messages to receive.
